File: CC++/Tree/BinaryTree.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node *left, *right;
    Node() {
        data = 0;
        left = right = nullptr;
    }
    Node(int _data) {
        data = _data;
        left = right = nullptr;
    }
};

struct BTree {
    Node* root;
    BTree() {
        root = nullptr;
    }

    void insert(int value) {
        if (root == nullptr)
            root = new Node(value);
        else {
            Node* itr = root;
            while (itr) {
                if (itr->data > value) {
                    if (itr->left) itr = itr->left;
                    else {
                        itr->left = new Node(value);
                        break;
                    }
                }
                else if (itr->data < value) {
                    if (itr->right) itr = itr->right;
                    else {
                        itr->right = new Node(value);
                        break;
                    }
                }
                else break;
            }
        }
    }

    void inOrder(Node* node) {
        if (node == nullptr) return ;
        inOrder(node->left);
        cout << node->data << " ";
        inOrder(node->right);
    }

    void preOrder(Node* node) {
        if (node == nullptr) return ;
        cout << node->data << " ";
        preOrder(node->left);
        preOrder(node->right);
    }

    Node* remove(Node* root, int k) {
        if (root == nullptr) return root;
        if (k < root->data) root->left = remove(root->left, k);
        else if (k > root->data) root->right = remove(root->right, k);
        else {
            if (root->left && root->right) {
                Node* replace = root->right;
                while (replace->left) replace = replace->left;
                root->data = replace->data;
                root->right = remove(root->right, root->data);
            }
            else if (root->left) {
                Node* temp = root->left;
                delete root;
                return temp;
            }
            else if (root->right) {
                Node* temp = root->right;
                delete root;
                return temp;
            }
            else {
                delete root;
                return nullptr;
            }
        }
        return root;
    }
};

int main() {
    int arr[] = {3, 2, 6, 2, 1, 5, 8};
    BTree t;
    for (int i : arr)
        t.insert(i);
    cout << "\nPre Order: \n";
    t.preOrder(t.root);
    cout << "\nIn Order: \n";
    t.inOrder(t.root);
    t.root = t.remove(t.root, 3);
    cout << "\nPre Order: \n";
    t.preOrder(t.root);
    cout << "\nIn Order: \n";
    t.inOrder(t.root);
}

File: binarytree/BinaryTree.java
public class BinaryTree {
    public class Node {
        int value;
        Node left, right;

        public Node(int value) {
            this.value = value;
        }
    }

    Node root;
    public BinaryTree() {
        root = null;
    }

    private Node insert(Node node, int value) {
        if (node == null)
            return new Node(value);
        if (value < node.value)
            node.left = insert(node.left, value);
        else if (value > node.value)
            node.right = insert(node.right, value);
        
        return node;
    }
    public void insert(int value) {
        root = insert(root, value);
    }

    private Node remove(Node node, int value) {
        if (node == null)
            return node;
        if (value < node.value)
            node.left = remove(node.left, value);
        else if (value > node.value)
            node.right = remove(node.right, value);
        else {
            // neu node khong co node con
            if (node.left == null && node.right == null)
                return null;
            // neu node co mot node con
            if (node.left == null)
                return node.right;
            if (node.right == null)
                return node.left;
            // neu node co hai node con
            // tim node trai nhat cua cay con ben phai 
            Node mostLeft = node.right;
            for (; mostLeft.left != null; mostLeft = mostLeft.left)
                ;
            // swap value of two nodes
            Util.swap(mostLeft, node);
            // de quy xoa node mostLeft
            node.right = remove(node.right, mostLeft.value);
        }

        return node;
    }
    public void remove(int value) {
        root = remove(root, value);
    }

    private void preOrder(Node node) {
        if (node == null) return;
        System.out.print(node.value + " ");
        preOrder(node.left);
        preOrder(node.right);
    }
    private void inOrder(Node node) {
        if (node != null) {
            inOrder(node.left);
            System.out.print(node.value + " ");
            inOrder(node.right);
        }
    }
    private void postOrder(Node node) {
        if (node != null) {
            postOrder(node.left);
            postOrder(node.right);
            System.out.print(node.value + " ");
        }
    }

    public static enum TraversalType {
        PREORDER, INORDER, POSTORDER
    }
    public void traverse(TraversalType type) {
        switch (type) {
            case PREORDER:
                preOrder(root);
                break;
            case INORDER:
                inOrder(root);
                break;
            case POSTORDER:
                postOrder(root);
                break;
            default:
                break;
        }
        System.out.println();
    }
}

class Util {
    public static void swap(BinaryTree.Node node1, BinaryTree.Node node2) {
        var temp = node1.value;
        node1.value = node2.value;
        node2.value = temp;
    }
}